# Provide a CMake option to select which CUDA architecture to build against.
# See first commit in https://github.com/espressomd/espresso/pull/4642
diff --git a/cmake/FindCUDACompilerNVCC.cmake b/cmake/FindCUDACompilerNVCC.cmake
--- a/cmake/FindCUDACompilerNVCC.cmake
+++ b/cmake/FindCUDACompilerNVCC.cmake
@@ -52,8 +52,16 @@
 list(APPEND CUDA_NVCC_FLAGS_RELWITHASSERT -O3 -g -Xptxas=-O3 -Xcompiler=-O3,-g)
-if(CMAKE_CUDA_COMPILER_VERSION VERSION_LESS 11)
-  list(APPEND CUDA_NVCC_FLAGS -gencode=arch=compute_30,code=sm_30)
+if(NOT DEFINED ESPRESSO_CUDA_ARCHITECTURES)
+  if("$ENV{CUDAARCHS}" STREQUAL "")
+    set(ESPRESSO_CUDA_ARCHITECTURES "75;86;89" CACHE INTERNAL "")
+  else()
+    set(ESPRESSO_CUDA_ARCHITECTURES "$ENV{CUDAARCHS}" CACHE INTERNAL "")
+  endif()
 endif()
+foreach(ESPRESSO_CUDA_ARCH ${ESPRESSO_CUDA_ARCHITECTURES})
+  list(APPEND CUDA_NVCC_FLAGS
+       "-gencode=arch=compute_${ESPRESSO_CUDA_ARCH},code=sm_${ESPRESSO_CUDA_ARCH}"
+       "-gencode=arch=compute_${ESPRESSO_CUDA_ARCH},code=compute_${ESPRESSO_CUDA_ARCH}")
+endforeach()
 list(APPEND CUDA_NVCC_FLAGS
-       -gencode=arch=compute_52,code=sm_52
-       -gencode=arch=compute_52,code=compute_52 -std=c++${CMAKE_CUDA_STANDARD}
+       -std=c++${CMAKE_CUDA_STANDARD}
        $<$<BOOL:${WARNINGS_ARE_ERRORS}>:-Xcompiler=-Werror;-Xptxas=-Werror>
